File: web/lib/league/live-stakes.ts
import { analyze, maximum, validRolls } from "../bowling";
import type { Night } from "../scorebook";
import { BOWLERS } from "../season";

export type LiveStake = {
  rosterIndex:number; name:string; game:number;
  title:string; detail:string; targetScratch?:number;
  status:"needs"|"clinch"|"tied"|"outOfReach"|"waiting";
};

/** Bounds include every outstanding strike/spare bonus, using legal completions. */
export function finalScoreBounds(rolls:number[]):{minimum:number;maximum:number;complete:boolean}|null {
  if(!validRolls(rolls))return null;
  const state=analyze(rolls), minimumRolls=[...rolls];
  while(!analyze(minimumRolls).complete)minimumRolls.push(0);
  return {minimum:analyze(minimumRolls).score,maximum:maximum(rolls),complete:state.complete};
}

/** Current individual matchups only. Pre-bowl scores are not live league matchups. */
export function liveStakes(night:Night):LiveStake[] {
  const match=night.match;
  if(!match || night.prebowl || night.game<1 || night.game>3)return [];
  const indices=match.ours.map(b=>BOWLERS.indexOf(b.name));
  // Never silently attribute an unknown/duplicated lineup name to another bowler.
  if(indices.length!==4 || indices.some(i=>i<0) || new Set(indices).size!==4)return [];
  const validHandicap=(n:unknown):n is number=>Number.isInteger(n)&&Number(n)>=0&&Number(n)<=120;
  const validScore=(n:unknown):n is number=>Number.isInteger(n)&&Number(n)>=0&&Number(n)<=300;
  return match.ours.flatMap((ours,slot):LiveStake[]=>{
    const rosterIndex=indices[slot], name=BOWLERS[rosterIndex], opponent=match.opponent.bowlers[slot];
    if(!opponent || !validHandicap(ours.handicap) || !validHandicap(opponent.handicap))return [];
    const context={rosterIndex,name,game:night.game};
    const other=match.opponentGames[night.game-1]?.[slot];
    if(other==null)return [{...context,status:"waiting",title:`${name}'s matchup is waiting`,detail:"Enter the opponent's final score to see the target."}];
    if(!validScore(other))return [];
    const override=night.finals?.[rosterIndex];
    if(override!=null && !validScore(override))return [];
    const bounds=override!=null?{minimum:override,maximum:override,complete:true}:finalScoreBounds(night.rolls[rosterIndex]);
    if(!bounds)return [];
    const tieScratch=other+opponent.handicap-ours.handicap;
    const targetScratch=Math.max(0,tieScratch+1);
    const base={...context,targetScratch};
    if(bounds.minimum>tieScratch)return [{...base,status:"clinch",title:bounds.complete?`${name} wins the matchup`:`${name} has clinched the matchup`,detail:bounds.complete?`${bounds.minimum} scratch, ${bounds.minimum+ours.handicap} with handicap.`:`Even a ${bounds.minimum} finish wins. Maximum: ${bounds.maximum}.`}];
    if(bounds.complete && bounds.minimum===tieScratch)return [{...base,status:"tied",title:`${name}'s matchup is tied`,detail:`${bounds.minimum+ours.handicap} each with handicap.`}];
    if(bounds.maximum<targetScratch)return [{...base,status:"outOfReach",title:bounds.complete?`${name}'s opponent wins`:`${name} cannot pass the opponent`,detail:bounds.maximum===tieScratch?`A ${bounds.maximum} finish can still tie the matchup.`:`Needs ${targetScratch} scratch to win. ${bounds.complete?"Finished":"Maximum"}: ${bounds.maximum}.`}];
    return [{...base,status:"needs",title:`${name} needs ${targetScratch} to win`,detail:`Scratch target, including the handicap difference. Maximum: ${bounds.maximum}.`}];
  });
}

File: ios/Sources/LiveStakes.swift
import Foundation

struct LiveStake: Codable, Equatable, Identifiable {
    enum Status: String, Codable { case needs, clinch, tied, outOfReach, waiting }
    var rosterIndex: Int
    var name: String
    var game: Int
    var title: String
    var detail: String
    var targetScratch: Int?
    var status: Status
    var id: String { "\(game)-\(rosterIndex)" }
}

enum LiveStakes {
    struct Bounds: Codable, Equatable { var minimum: Int; var maximum: Int; var complete: Bool }
    static func finalScoreBounds(_ rolls: [Int]) -> Bounds? {
        var game = BowlingGame()
        guard rolls.count <= 21, rolls.allSatisfy({ game.add($0) }) else { return nil }
        var minimum = game
        while !minimum.isComplete { minimum.add(0) }
        return Bounds(minimum: minimum.settledScore, maximum: game.maximumScore, complete: game.isComplete)
    }

    static func cards(for night: Night) -> [LiveStake] {
        guard let match = night.match, night.prebowl == nil, (1...3).contains(night.game), night.rolls.count == 4,
              night.finals == nil || night.finals?.count == 4 else { return [] }
        let indices = match.ours.compactMap { Night.names.firstIndex(of: $0.name) }
        guard match.ours.count == 4, indices.count == 4, Set(indices).count == 4 else { return [] }
        return match.ours.enumerated().compactMap { slot, ours in
            guard match.opponent.bowlers.indices.contains(slot) else { return nil }
            let opponent = match.opponent.bowlers[slot], index = indices[slot], name = Night.names[index]
            guard (0...120).contains(ours.handicap), (0...120).contains(opponent.handicap) else { return nil }
            func card(_ status: LiveStake.Status, _ title: String, _ detail: String, _ target: Int? = nil) -> LiveStake {
                LiveStake(rosterIndex: index, name: name, game: night.game, title: title, detail: detail, targetScratch: target, status: status)
            }
            let opponents = match.opponentGames.indices.contains(night.game - 1) ? match.opponentGames[night.game - 1] : []
            guard opponents.indices.contains(slot), let other = opponents[slot] else {
                return card(.waiting, "\(name)'s matchup is waiting", "Enter the opponent's final score to see the target.")
            }
            guard (0...300).contains(other) else { return nil }
            let bounds: Bounds
            if let score = night.finals?[index] {
                guard (0...300).contains(score) else { return nil }
                bounds = Bounds(minimum: score, maximum: score, complete: true)
            } else {
                guard let value = finalScoreBounds(night.rolls[index]) else { return nil }
                bounds = value
            }
            let tie = other + opponent.handicap - ours.handicap, target = max(0, tie + 1)
            if bounds.minimum > tie {
                return card(.clinch, bounds.complete ? "\(name) wins the matchup" : "\(name) has clinched the matchup", bounds.complete ? "\(bounds.minimum) scratch, \(bounds.minimum + ours.handicap) with handicap." : "Even a \(bounds.minimum) finish wins. Maximum: \(bounds.maximum).", target)
            }
            if bounds.complete && bounds.minimum == tie {
                return card(.tied, "\(name)'s matchup is tied", "\(bounds.minimum + ours.handicap) each with handicap.", target)
            }
            if bounds.maximum < target {
                let detail = bounds.maximum == tie ? "A \(bounds.maximum) finish can still tie the matchup." : "Needs \(target) scratch to win. \(bounds.complete ? "Finished" : "Maximum"): \(bounds.maximum)."
                return card(.outOfReach, bounds.complete ? "\(name)'s opponent wins" : "\(name) cannot pass the opponent", detail, target)
            }
            return card(.needs, "\(name) needs \(target) to win", "Scratch target, including the handicap difference. Maximum: \(bounds.maximum).", target)
        }
    }
}
